compiler/rustc_const_eval/src/const_eval/valtrees.rs

```diff
@@ -0,0 +1,479 @@
+                &(*place).into(),
+            )
+            .unwrap();
+        }
+        ty::Ref(_, inner_ty, _) => {
+            match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
+                    let slice_ty = match inner_ty.kind() {
+                        ty::Slice(slice_ty) => *slice_ty,
+                        ty::Str => tcx.mk_ty(ty::Uint(ty::UintTy::U8)),
+                        _ => bug!("expected ty::Slice | ty::Str"),
+                    };
+                    debug!(?slice_ty);
+
+                    let valtrees = valtree.unwrap_branch();
+                    debug!(?valtrees);
+                    let len = valtrees.len();
+                    debug!(?len);
+
+                    // create a place for the underlying array
+                    let arr_ty = tcx.mk_array(slice_ty, len as u64);
+                    let mut arr_place =
+                        create_mplace_from_layout(ecx, ty::ParamEnv::empty().and(arr_ty));
+                    debug!(?arr_place);
+
+                    // Insert elements of `arr_valtree` into `place`
+                    fill_place_recursively(ecx, &mut arr_place, valtree, arr_ty);
+                    dump_place(&ecx, arr_place.into());
+
+                    // Now we need to create a `ScalarPair` from the filled `place`
+                    // and write that into `place`
+                    let (alloc_id, offset) = arr_place.mplace.ptr.into_parts();
+                    debug!(?alloc_id, ?offset);
+                    let unwrapped_ptr = Pointer { offset, provenance: alloc_id.unwrap() };
+                    let len_scalar = ScalarMaybeUninit::Scalar(Scalar::from_u64(len as u64));
+
+                    let imm = Immediate::ScalarPair(
+                        ScalarMaybeUninit::from_pointer(unwrapped_ptr, &tcx),
+                        len_scalar,
+                    );
+                    debug!(?imm);
+
+                    // Now write the ScalarPair into the original place we wanted to fill
+                    // in this call
+                    let _ = ecx.write_immediate(imm, &(*place).into()).unwrap();
+
+                    dump_place(&ecx, (*place).into());
+                }
+                _ => {
+                    let mut pointee_place =
+                        create_mplace_from_layout(ecx, ty::ParamEnv::empty().and(*inner_ty));
+                    debug!(?pointee_place);
+                    fill_place_recursively(ecx, &mut pointee_place, valtree, *inner_ty);
+
+                    dump_place(ecx, pointee_place.into());
+                    intern_const_alloc_recursive(ecx, InternKind::Constant, &pointee_place)
+                        .unwrap();
+
+                    let imm = pointee_place.mplace.to_ref(&tcx);
+                    debug!(?imm);
+
+                    ecx.write_immediate(imm, &(*place).into()).unwrap();
+                }
+            }
+        }
+        ty::Tuple(tuple_types) => {
```

Comment:
This and all the arms below can use the same logic (with just an adjustment to skip the variant logic and downcast for non-enums).

The field types are already in the mplace from the field projection, so you can ignore the types and not deconstruct them yourself